elki/src/main/java/de/lmu/ifi/dbs/elki/math/statistics/dependence/McdeMwpDependenceMeasure.java

```diff
@@ -0,0 +1,227 @@
+import de.lmu.ifi.dbs.elki.utils.containers.MwpIndex;
+
+/**
+ * Implementation of bivariate Monte Carlo Density Estimation using Mann-Withney U test, known as MWP. See
+ * Edouard Fouché & Klemens Böhm<br>
+ * Monte Carlo Density Estimation<br>
+ * Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)
+ *
+ * This class extends MCDEDependenceMeasure and implements the Mann-Whitney-U statistical test and an appropriate index structure.
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+        title = "Monte Carlo Density Estimation", //
+        booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)",
+        url = "https://doi.org/10.1145/3335783.3335795", //
+        bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+public class McdeMwpDependenceMeasure extends MCDEDependenceMeasure<MwpIndex> {
+
+    /**
+     * Constructor
+     *
+     * @param m Monte-Carlo iterations
+     * @param alpha Expected share of instances in slice (under independence)
+     * @param beta Share of instances in marginal restriction (reference dimension)
+     * @param rnd Random source
+     */
+    public McdeMwpDependenceMeasure(int m, double alpha, double beta, RandomFactory rnd){
+        super(m, alpha, beta, rnd);
+    }
+
+    /**
+     * Computes Corrected Rank Index as described in Algorithm 1 of reference paper.
+     * The notation follows ELKI convention if applicable, else we use the notation from the reference paper.
+     *
+     * @param adapter ELKI NumberArrayAdapter Subclass
+     * @param data One dimensional array containing one dimension of the data
+     * @param idx Return value of sortedIndex()
+     * @return Array of doubles, 3 subsequent values being assigned to one data instance.
+     * Containing sorted (ascending) row numbers, adjusted ranks and tying value corrections
+     * as required by MWP test. Example:
+     * double[] corrected_ranks = corrected_ranks(...);
+     * double l = corrected_rank[0]; double adjusted_rank = corrected_rank[1]; double correction = corrected_rank[2];
+     * 
+     * Correspond to one instance of the original data
+     */
+    protected <A> MwpIndex[] corrected_ranks(final NumberArrayAdapter<?, A> adapter, final A data, int[] idx){
+        final int len = adapter.size(data);
+        MwpIndex[] I = (MwpIndex[]) Array.newInstance(MwpIndex.class, len);
+
+        int j = 0; int correction = 0;
+        while(j < len){
+            int k = j; int t = 1; double adjust = 0.0;
+
+            while((k < len - 1) && (adapter.getDouble(data, idx[k]) == adapter.getDouble(data, idx[k+1]))){
+                adjust += k;
+                k++; t++;
+            }
+
+            if(k > j){
+                double adjusted = (adjust + k) / t;
+                correction += (t*t*t) - t;
```

Comment:
As far as I can tell, `correction` is always integer. These values here likely can get quite large - wouldn't a `long` be more appropriate? Also, the data is later stored as a `double`?